conjure_oxide/src/parse.rs

```diff
@@ -124,12 +133,87 @@ fn parse_int_domain(v: &JsonValue) -> Result<Domain> {
     Ok(Domain::IntDomain(ranges))
 }
 
-fn parse_constraint(_obj: &JsonValue) -> Result<()> {
-    Ok(())
+fn parse_expression(obj: &JsonValue) -> Option<Expression> {
+    // this needs an explicit type signature to force the closures to have the same type
+    type BinOp = Box<dyn Fn(Box<Expression>, Box<Expression>) -> Expression>;
+    let binary_operators: HashMap<&str, BinOp> = [
+        ("MkOpEq", Box::new(Expression::Eq) as Box<dyn Fn(_, _) -> _>),
+        (
+            "MkOpNeq",
+            Box::new(Expression::Neq) as Box<dyn Fn(_, _) -> _>,
+        ),
+        (
+            "MkOpGeq",
+            Box::new(Expression::Geq) as Box<dyn Fn(_, _) -> _>,
+        ),
+        (
+            "MkOpLeq",
+            Box::new(Expression::Leq) as Box<dyn Fn(_, _) -> _>,
+        ),
+        ("MkOpGt", Box::new(Expression::Gt) as Box<dyn Fn(_, _) -> _>),
+        ("MkOpLt", Box::new(Expression::Lt) as Box<dyn Fn(_, _) -> _>),
+    ]
+    .into_iter()
+    .collect();
+
+    let mut binary_operator_names = binary_operators.iter().map(|x| x.0);
+
+    match obj {
+        Value::Object(op) if op.contains_key("Op") => {
+            match &op["Op"] {
+                Value::Object(bin_op)
+                    if binary_operator_names.any(|key| bin_op.contains_key(*key)) =>
+                {
+                    // we know there is a single key value pair in this object
+                    // extract the value, ignore the key
+                    let (key, value) = bin_op.into_iter().next()?;
+
+                    let constructor = binary_operators.get(key.as_str())?;
+
+                    match &value {
+                        Value::Array(bin_op_args) if bin_op_args.len() == 2 => {
+                            let arg1 = parse_expression(&bin_op_args[0])?;
+                            let arg2 = parse_expression(&bin_op_args[1])?;
+                            Some(constructor(Box::new(arg1), Box::new(arg2)))
+                        }
+                        otherwise => panic!("Unhandled 3 {:#?}", otherwise),
+                    }
+                }
+                Value::Object(op_sum) if op_sum.contains_key("MkOpSum") => {
+                    let args = &op_sum["MkOpSum"]["AbstractLiteral"]["AbsLitMatrix"][1];
+                    let args_parsed: Vec<Expression> = args
+                        .as_array()?
+                        .iter()
+                        .map(|x| parse_expression(x).unwrap())
+                        .collect();
+                    Some(Expression::Sum(args_parsed))
+                }
+                otherwise => panic!("Unhandled 2 {:#?}", otherwise),
+            }
+        }
+        Value::Object(refe) if refe.contains_key("Reference") => {
```

Comment:
I imagine this function is going to get bigger? might be worth setting a precident by pulling out things like this into a `parse_reference` function, below into `parse_constant`, so this function doesn't get too huge to read (it's mostly just a case statement that dispatches).